src/client/line.cpp

```diff
@@ -0,0 +1,547 @@
+		{
+			v2u32 frame_size;
+			m_params.animation.determineParams(tex_size, nullptr, nullptr, &frame_size);
+
+			f32 aspect = (f32)frame_size.Y / frame_size.X;
+			f32 tcoords_u = dist * aspect / m_params.width;
+
+
+			ll_corner_coords = m_params.animation.getTextureCoords(tex_size, m_current_animation_frame);
+			ur_corner_coords.X = ll_corner_coords.X + (f32)frame_size.X/tex_size.Width*tcoords_u;
+			ur_corner_coords.Y = ll_corner_coords.Y + (f32)frame_size.Y/tex_size.Height;
+		}
+		else
+		{
+			f32 aspect = (f32)tex_size.Height / tex_size.Width;
+			f32 tcoords_u = dist * aspect / m_params.width;
+
+			ur_corner_coords.X = tcoords_u;
+		}
+	}
+	video::S3DVertex vertices[4] = {
+		video::S3DVertex(positions[0], line_plane.Normal, m_color.first, v2f(ll_corner_coords.X, ur_corner_coords.Y)),
+		video::S3DVertex(positions[1], line_plane.Normal, m_color.first, ll_corner_coords),
+		video::S3DVertex(positions[2], line_plane.Normal, m_color.second, v2f(ur_corner_coords.X, ll_corner_coords.Y)),
+		video::S3DVertex(positions[3], line_plane.Normal, m_color.second, ur_corner_coords)
+	};
+
+	video::IVideoDriver* vdrv = SceneManager->getVideoDriver();
+
+	vdrv->setTransform(video::ETS_WORLD, core::IdentityMatrix);
+	vdrv->setMaterial(m_mat);
+
+	u16 indices[] = {0, 1, 2, 2, 3, 0};
+
+	vdrv->drawIndexedTriangleList(vertices, 4, indices, 2);
+}
+
+void LineSceneNode::drawSpriteStrip() const
+{
+	video::IVideoDriver* vdrv = SceneManager->getVideoDriver();
+
+	vdrv->setTransform(video::ETS_WORLD, core::IdentityMatrix);
+	vdrv->setMaterial(m_mat);
+
+	vdrv->draw3DLine(m_pos.first, m_pos.second, m_color.first);
+}
+
+void LineSceneNode::calc3DCircle(bool first, video::S3DVertex* verts, u16* inds, u32 verts_c) const
+{
+	const v3f& center_pos = first ? m_pos.first : m_pos.second;
+	v3f normal(0.0f, 0.0f, first ? -1.0f : 1.0f);
+	video::SColor color = first ? m_color.first : m_color.second;
+	v2f tcoords(0.5f, 0.5f);
+
+	verts[0] = video::S3DVertex(center_pos, normal, color, tcoords);
+	tcoords.X = 0.0f;
+
+	v3f radius(-m_params.width/2, 0.0f, 0.0f);
+
+	f32 rot_steps = 360.0f / (verts_c-1);
+
+	inds[0] = 0;
+
+	core::matrix4 trans_mat;
+	trans_mat.setTranslation(center_pos);
+
+	v3f dist = m_pos.second - m_pos.first;
+
+	core::matrix4 rot_mat;
+	rot_mat.buildRotateFromTo(core::vector3df(0.0f, 0.0f, 1.0f), dist);
+
+	core::matrix4 res_mat = trans_mat * rot_mat;
+
+	for (u32 i = 0; i < verts_c-1; i++)
+	{
+		v3f rotated_radius = radius;
+		rotated_radius.rotateXYBy(-rot_steps * i);
+		res_mat.transformVect(rotated_radius);
+
+		v2f rotated_tcoords = tcoords;
+		rotated_tcoords.rotateBy(rot_steps * i);
+		rotated_tcoords -= v2f(-0.5f, -0.5f);
+
+		verts[i+1] = video::S3DVertex(
+			rotated_radius,
+			normal,
+			color,
+			rotated_tcoords
+		);
+
+		inds[i+1] = first ? i + 1 : (i == 0 ? 1 : verts_c - i);
+	}
+	inds[verts_c] = 1;
+}
+
+void LineSceneNode::drawCylindricShape() const
+{
+	const u32 circle_verts_c = 21; // including the center
```

Comment:
Because the circle has 20 vertices (I selected such count to give it a look symmetrical, not too polygonal and minimal as possible) plus still two vertices one of which is closed with other one them to have various texture coordinates and another one is a center.